questionpy_server/collector/local_collector.py

```diff
@@ -201,32 +141,103 @@ def __init__(self, directory: Path, indexer: 'Indexer'):
+        """
+        Reflect changes in the directory to the indexer and internal map.
+
+        :param with_log: Whether to log the changes.
+        """
+
+        def directory_iterator(directory: str) -> Generator[Path, Any, None]:
+            """
+            Iterates over all packages in the directory.
+            Used as the custom directory iterator for DirectorySnapshot.
+
+            :param directory: The directory.
+            :return: A generator of paths.
+            """
+
+            for file in Path(directory).glob('*.qpy'):
+                if file.is_file():
+                    yield file
+
+        async def add_package(pkg_hash: str, pkg_path: Path) -> None:
+            """
+            Adds a package to the map and registers it in the indexer.
+
+            :param pkg_hash: The hash of the package.
+            :param pkg_path: The path of the package.
+            """
+
+            self.map.insert(pkg_hash, pkg_path)
+            await self.indexer.register_package(pkg_hash, pkg_path, self)
+
+        async def remove_package(pkg_path: Path) -> None:
+            """
+            Removes a package from the map and unregisters it from the indexer.
+
+            :param pkg_path: The path of the package.
+            """
+
+            pkg_hash = self.map.pop(pkg_path)
+            if not pkg_hash:
+                return
+            packages = self.map.get(pkg_hash)
+            if packages is None or len(packages) == 0:
+                # There is no other package with the same hash - unregister it.
+                await self.indexer.unregister_package(pkg_hash, self)
+
+        # If no snapshot exists, use EmptyDirectorySnapshot to get all files as created.
+        old_snapshot = self._snapshot or EmptyDirectorySnapshot()
+        new_snapshot = DirectorySnapshot(str(self.directory), recursive=False, listdir=directory_iterator)
+        difference = DirectorySnapshotDiff(old_snapshot, new_snapshot)
+
+        for path in difference.files_created:
+            package_path = Path(path)
+            package_hash = await to_thread(calculate_hash, package_path)
+            await add_package(package_hash, package_path)
+
+        for path in difference.files_deleted:
+            package_path = Path(path)
+            await remove_package(package_path)
+
+        for path in difference.files_modified:
+            package_path = Path(path)
+            package_hash = await to_thread(calculate_hash, package_path)
+            await remove_package(package_path)
+            await add_package(package_hash, package_path)
+
+            self._log.warning("Package %s was modified. This will cause unexpected behavior if the package is "
+                              "currently read by a worker.", package_path)
+
+        # Replace paths for moved packages.
+        paths = [(Path(old_path), Path(new_path)) for old_path, new_path in difference.files_moved]
+        self.map.replace(paths)
```

Comment:
Ich musste die Funktion `PathToHash.replace` mehrmals lesen. Auch wenn es korrekt aussieht, fände ich es besser, wenn du über `difference.files_moved` einfach genauso iterieren würdest wie in den anderen drei Fällen. Ich glaube, dass es dann einfacher verständlich ist.

Reply:
Ich habe den Code angepasst. Die Idee war es, die `replace`-Logik in `PathToHash` zu platzieren, um `update` nicht noch länger (und damit evtl. unübersichtlich) zu machen.